### PR TITLE
30 service unit testing

### PR DESCRIPTION
**Pull request by zirka28**

### CODE REVIEW

File: moninag/tests/servise_model_test.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+from django.test import TestCase
+from service.models import Service
+from server.models import Server
+from registration.models import CustomUser
+
+class ServiceTest(TestCase):
```
Comment (ghost) [inline moninag/tests/servise_model_test.py]: expected 2 blank lines, found 1

File: moninag/tests/servise_model_test.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    def setUp(self):
+        user=CustomUser.objects.create(id=1, first_name="firstname", second_name="secondname", email="email@gmail.com")
```
Comment (ghost) [inline moninag/tests/servise_model_test.py]: missing whitespace around operator

Comment (ghost) [reply moninag/tests/servise_model_test.py]: line too long (119 > 79 characters)

File: moninag/tests/servise_model_test.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        user1=CustomUser.objects.create(id=2, first_name="firstname2", second_name="secondname2", email="email2@gmail.com")
+        user2=CustomUser.objects.create(id=3, first_name="firstname3", second_name="secondname3", email="email3@gmail.com")
+        server=Server.objects.create(id =1, name="name", address="address", state="Production", user=user)
```
Comment (ghost) [inline moninag/tests/servise_model_test.py]: unexpected spaces around keyword / parameter equals

File: moninag/tests/servise_model_test.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        ser1.update(name="service2")
+        self.assertEqual(ser1.name, "service2")
+
+    def test_service_create(self):
+        server = Server.objects.get(id=1)
+        Service.create("service_new", server, "ok")
+        service = Service.objects.get(name="service_new")
+        
+        self.assertEqual(service.name,"service_new")
+        self.assertEqual(service.status,"ok")
```
Comment (ghost) [inline moninag/tests/servise_model_test.py]: missing whitespace after ','

File: moninag/tests/servise_model_test.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        self.assertEqual(service.server, server)
+
+
+    def test_service_get_by_id(self):
```
Comment (ghost) [inline moninag/tests/servise_model_test.py]: too many blank lines (2)

File: moninag/tests/unittests/service/test_service_model.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
+from django.test import TestCase
+from service.models import Service
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_model.py]: blank line between project and third party imports

File: moninag/tests/unittests/service/test_service_model.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+from server.models import Server
+from registration.models import CustomUser
+
+class ServiceTest(TestCase):
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_model.py]: expected 2 blank lines, found 1

File: moninag/tests/unittests/service/test_service_model.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    def setUp(self):
+        user=CustomUser.objects.create(id=1, first_name="firstname", second_name="secondname", email="email@gmail.com")
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_model.py]: missing whitespace around operator

Comment (ghost) [reply moninag/tests/unittests/service/test_service_model.py]: line too long (119 > 100 characters)

File: moninag/tests/unittests/service/test_service_model.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        user1=CustomUser.objects.create(id=2, first_name="firstname2", second_name="secondname2", email="email2@gmail.com")
+        user2=CustomUser.objects.create(id=3, first_name="firstname3", second_name="secondname3", email="email3@gmail.com")
+        server=Server.objects.create(id =1, name="name", address="address", state="Production", user=user)
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_model.py]: unexpected spaces around keyword / parameter equals

File: moninag/tests/unittests/service/test_service_model.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        self.assertEqual(ser1.name, "service2")
+
+    def test_service_create(self):
+        server = Server.objects.get(id=1)
+        Service.create("service_new", server, "ok")
+        service = Service.objects.get(name="service_new")
+        
+        self.assertEqual(service.name,"service_new")
+        self.assertEqual(service.status,"ok")
+        self.assertEqual(service.server, server)
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_model.py]: missing whitespace after ','

File: moninag/tests/unittests/service/test_service_view.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+        self.assertEqual(actual_response.status_code, 200)
+        self.assertEqual(service.name, 'Service101')
+        self.assertEqual(service.status, 'fail')
+
+    def test_put_invalid_format(self):
+
+        url = reverse('service', args=[11])
+        data = json.dumps({'name': 'Service101',
+                           'status': '15',
+                           'user_id':18 })
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_view.py]: whitespace before '}'

File: moninag/tests/unittests/service/test_service_view.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+            id=1,
+            first_name='Name',
+            second_name='Surname',
+            email='email@gmail.com',
+            is_active=True,
+        )
+
+        self.user.set_password('qwerty')
+        self.user.save()
+        self.user = authenticate(username='email@gmail.com', password='qwerty')
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_view.py]: actually you dont need this line 😄

File: moninag/tests/unittests/service/test_service_view.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    def test_put_not_existing_service(self):
+
+        url = reverse('service', args=[44])
+        data = json.dumps({'name': 'Service101',
+                           'status': 'ok'})
+
+        actual_response = self.client.put(url, data=data, content_type='application/json')
+
+        self.assertEqual(actual_response.status_code, 404)
+    def test_put_not_user_service(self):
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_view.py]: expected 1 blank line, found 0

**Review by ghost — CHANGES_REQUESTED** [review]

File: moninag/tests/unittests/service/test_service_view.py [hunk 14]
```diff
@@ -0,0 +1,238 @@
+from django.test import TestCase
+from django.test import Client
+from django.core.urlresolvers import reverse
+from django.contrib.auth import authenticate
+import json
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_view.py]: reorder imports according to pep8

File: moninag/tests/unittests/service/test_service_model.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+        server2 = Server.objects.create(id=2,
+                                        name="name2",
+                                        address="address2",
+                                        state="Production",
+                                        user=user2)
+
+        Service.objects.create(id=20,
+                               name="service1",
+                               status="ok", server=server2)
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_model.py]: newline before server=server2

File: moninag/tests/unittests/service/test_service_model.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+                                          first_name="firstname2",
+                                          second_name="secondname2",
+                                          email="email2@gmail.com")
+
+        user2 = CustomUser.objects.create(id=3,
+                                          first_name="firstname3",
+                                          second_name="secondname3",
+                                          email="email3@gmail.com")
+
+        server = Server.objects.create(id=1,
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_model.py:20]: unused variable **server**, probably you don't need this assignment

File: moninag/tests/unittests/service/test_service_model.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+
+class ServiceTest(TestCase):
+    def setUp(self):
+        user = CustomUser.objects.create(id=1,
+                                         first_name="firstname",
+                                         second_name="secondname",
+                                         email="email@gmail.com")
+
+        user1 = CustomUser.objects.create(id=2,
```
Comment (ghost) [inline moninag/tests/unittests/service/test_service_model.py]: unused variable **user1**, probably you don't need this assignment

File: moninag/tests/unittests/service/test_service_model.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    def test_service_get_by_server_id(self):
+        server = Server.objects.get(id=1)
+        actual = Service.objects.filter(server=server)
+        expect = Service.get_by_server_id(1)
+        self.assertQuerysetEqual(expect, map(repr, actual), ordered=False)
+
+    def test_service_get_by_user_id(self):
+        user = CustomUser.objects.get(id=3)
+        servers = Server.objects.filter(user=user)
+        actual = Service.objects.filter(server__in=servers)
```
Comment (hutsulyk) [inline moninag/tests/unittests/service/test_service_model.py]: You could simplify this :slightly_smiling_face: 
`services = Service.objects.filter(server__user__id=3)`

File: moninag/tests/unittests/service/test_service_view.py [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+        servers = Server.objects.filter(user=1)
+        services = Service.objects.filter(server__in=servers)
+
+        expected_response = {}
+        expected_response['response'] = [to_dict(service)
+                                         for service in services]
+
+        self.assertEqual(actual_response.status_code, 200)
+        self.assertJSONEqual(actual_response.content.decode('utf-8'),
+                             expected_response)
```
Comment (hutsulyk) [inline moninag/tests/unittests/service/test_service_view.py]: Comparing `str` to `dict` :upside_down_face:

File: moninag/tests/unittests/service/test_service_view.py [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+                           'server_id': 10})
+
+        actual_response = self.client.post(url, data=data,
+                                           content_type='application/json')
+        service = Service.objects.get(name='Service')
+
+        self.assertEqual(actual_response.status_code, 201)
+        self.assertEqual(service.name, 'Service')
+        self.assertEqual(service.status, 'ok')
+        self.assertEqual(service.server.id, 10)
```
Comment (hutsulyk) [inline moninag/tests/unittests/service/test_service_view.py]: It is better to check `actual_response` JSON :slightly_smiling_face:

File: moninag/tests/unittests/service/test_service_view.py [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+        data = json.dumps({'name': 'Service101',
+                           'status': 'fail'})
+
+        actual_response = self.client.put(url, data=data,
+                                          content_type='application/json')
+        service = Service.objects.get(id=11)
+
+        self.assertEqual(actual_response.status_code, 200)
+        self.assertEqual(service.name, 'Service101')
+        self.assertEqual(service.status, 'fail')
```
Comment (hutsulyk) [inline moninag/tests/unittests/service/test_service_view.py]: It is better to check `actual_response` JSON :upside_down_face:

File: moninag/tests/unittests/service/test_service_model.py (hunk 16 above)
Comment (zirka28) [reply moninag/tests/unittests/service/test_service_model.py:20]: Now I use it :)

**Comment by hutsulyk** [comment]
`__init__.py` is missing :upside_down_face: 

Everything else works and looks good!

**Review by ghost — CHANGES_REQUESTED** [review]
I guess you should add docstrings for each test function since tests in dev already have them and it will help to avoid additional work when we will integrate with Jenkins

**Review by hutsulyk — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**